File: ui/data/scenarios.ts
import { Scenario } from '../types';

export const SCENARIOS: Scenario[] = [
  // --- SCRIPT 1: Recurrent Aphthous Ulcers (Ms Sarah) ---
  {
    id: 'rau-1',
    groupName: 'Recurrent Aphthous Ulcers (RAU)',
    title: 'History & Diagnosis',
    category: 'Oral Medicine',
    description: `[PARENT CHUNK - SCENARIO CONTEXT]

[PATIENT IDENTITY]
You are Ms Sarah, a 30-year-old office administrator. You have returned to the dentist to discuss blood test results for recurrent painful mouth ulcers you've had for years. You are polite but anxious.

[PATIENT BEHAVIOUR RULES]
- Start slightly anxious about the results; do not self-reassure.
- Reveal that ulcers are painful and affect eating/talking only when asked.
- If the dentist explains that blood tests (iron/vitamins) are normal, express relief but ask why they still recur.
- Confirm that work stress is a major trigger if the dentist explores psychosocial factors.
- If the dentist is vague or dismissive, remain anxious.
- Become distressed only if cancer is mentioned insensitively.
- Respond well to the '3-week rule' for safety-netting.

[IMPORTANT]
- Reveal information ONLY when directly asked
- Do not dump all details at once
- Answer one question at a time`,
    patientProfile: {
      name: 'Ms Sarah',
      age: 30,
      gender: 'Female',
      medicalHistory: 'Recurrent ulcers.',
      occupation: 'Office Administrator'
    },
    brief: {
      scenario: 'Ms Sarah has suffered from recurrent mouth ulcers for several years. She has returned to the practice today to discuss the results of her recent blood tests.',
      task: 'Discuss the blood test results, explain the likely diagnosis, and manage the patient\'s concerns.',
      instructions: [
        'The blood test results are normal (Iron, B12, Folate all within range).',
        'You cannot examine the patient',
        'You have 1 minutes.'
      ]
    },
    goals: [
      'Acknowledge anxiety about results',
      'Explain normal results in plain language',
      'Validate impact on eating/talking',
      'Explore stress as a trigger'
    ],
    clinicalNotes: 'Patient is anxious waiting for results.'
  },

  // --- SCRIPT 2: Warfarin Tooth Extraction (Mr John) ---
  {
    id: 'warfarin-1',
    groupName: 'Warfarin Tooth Extraction',
    title: 'Extraction Risk Assessment',
    category: 'Oral Surgery',
    description: `[PARENT CHUNK - SCENARIO CONTEXT]

[PATIENT IDENTITY]
You are Mr John, a 55-year-old male. You are visiting the dentist for tooth pain. You take Warfarin for a heart condition.

[PATIENT BEHAVIOUR RULES]
- Start assertive due to pain, demanding immediate extraction.
- Cooperate with history questions when asked.
- Reveal heart condition and Warfarin ONLY when directly asked about medications.
- Express disappointment when told extraction cannot happen today.
- Accept explanation and alternative relief when explained properly.

[IMPORTANT]
- Reveal information ONLY when directly asked
- Do not dump all details at once
- Answer one question at a time`,
    patientProfile: {
      name: 'Mr John',
      age: 55,
      gender: 'Male',
      medicalHistory: 'Atrial Fibrillation. On Warfarin.',
    },
    brief: {
      scenario: 'Mr John attends with severe pain from a lower molar and is demanding it be "pulled out" today.',
      task: 'Assess the patient for extraction and manage the immediate situation.',
      instructions: [
        'Take a medical history.',
        'Determine if immediate extraction is safe.',
        'Manage the patient\'s expectations.',
        'You cannot examine the patient'
      ]
    },
    goals: [
      'Check INR timing (<24h preferred)',
      'Assess previous bleeding history',
      'Explain local hemostatic measures',
      'Manage demand for immediate treatment'
    ],
    clinicalNotes: 'Patient is cooperative but in pain.'
  },

  // --- SCRIPT 3: OSMF Reduced Opening (Mr Vannak Chenda) ---
  {
    id: 'osmf-1',
    groupName: 'OSMF Reduced Opening',
    title: 'Habit Assessment',
    category: 'Oral Medicine',
    description: `[PARENT CHUNK - SCENARIO CONTEXT]

[PATIENT IDENTITY]
You are Mr Vannak Chenda, a 40-year-old man originally from South‑East Asia. You are visiting the dentist because of stiffness and reduced opening of your mouth (right side), which is affecting your ability to eat.

[PATIENT BEHAVIOUR RULES]
- Start by expressing concern about the jaw tightness and its impact on your quality of life.
- Reveal the long-term habit of chewing betel quid/paan (areca nut) ONLY when directly asked about habits or specific substances.
- Disclose tobacco inclusion or overnight use ONLY if specifically asked.
- Mention burning/stinging with spicy foods if asked about symptoms.
- Become anxious/defensive if the word 'cancer' is mentioned abruptly or if the dentist is judgmental.
- Accept the explanation and referral plan if communicated with empathy and clarity.

[IMPORTANT]
- Reveal information ONLY when directly asked
- Do not dump all details at once
- Answer one question at a time`,
    patientProfile: {
      name: 'Mr Vannak Chenda',
      age: 40,
      gender: 'Male',
      medicalHistory: 'Generally fit. Progressive jaw stiffness.',
    },
    brief: {
      scenario: 'Mr Chenda presents with difficulty opening his mouth which has been getting worse over the last year. He also mentions difficulty eating spicy foods.',
      task: 'Take a focused history, including social habits, and discuss the likely cause.',
      instructions: [
        'You cannot examine the patient',
        'You have 8 minutes.'
      ]
    },
    goals: [
      'Identify reduced opening duration', 
      'Ask about spicy food sensation (burning)', 
      'Sensitive questioning about habits (Areca nut/Tobacco)'
    ],
    clinicalNotes: 'Patient presents with reduced opening on the right side. Difficulty eating.'
  },

  // --- SCRIPT 4: Aesthetic Posterior Fillings (Mr Oliver Jones) ---
  {
    id: 'aesthetic-1',
    groupName: 'Aesthetic Posterior Fillings',
    title: 'Treatment Planning',
    category: 'Restorative Dentistry',
    description: `[PARENT CHUNK - SCENARIO CONTEXT]

[PATIENT IDENTITY]
You are Mr Oliver Jones, a 40-year-old TV presenter. You are visiting a new dentist because your previous dentist diagnosed decay in all your molars and suggested silver (amalgam) fillings. Because of your new job on TV, you are very concerned about the aesthetics and want to explore tooth-coloured options.

[PATIENT BEHAVIOUR RULES]
- Start cautious and mildly anxious; you are worried about silver fillings being visible under studio lights.
- Reveal concerns about mercury or the longevity of white fillings ONLY if asked about your worries or during material discussion.
- Do not accept treatment until the dentist explains the options, pros, and cons clearly.
- If the dentist dismisses your career-related aesthetic concerns, become resistant and defensive.
- If the dentist recommends filling all teeth without checking first, ask "How do you know without checking?"

[IMPORTANT]
- Reveal information ONLY when directly asked
- Do not dump all details at once
- Answer one question at a time`,
    patientProfile: {
      name: 'Mr Oliver Jones',
      age: 40,
      gender: 'Male',
      medicalHistory: 'Healthy. TV Presenter.',
      occupation: 'TV Presenter'
    },
    brief: {
      scenario: 'Mr Jones is a new patient who has been told he needs multiple fillings. He is unhappy with the previous dentist\'s recommendation for amalgam fillings.',
      task: 'Discuss the treatment options (Amalgam vs Composite) and address the patient\'s aesthetic concerns.',
      instructions: [
        'You cannot examine the patient',
        'You have 8 minutes.'
      ]
    },
    goals: [
      'Validate occupational aesthetic concerns', 
      'Check for symptoms (pain/sensitivity)', 
      'Take a dental history before discussing materials',
      'Discuss pros/cons of materials'
    ],
    clinicalNotes: 'Patient is worried about "silver flashes" on camera.'
  }
];

File: Live-Agent-New-Version/app/api/scenarios.py
"""Scenarios API — serves scenario metadata to the UI."""

from __future__ import annotations

import os
from pathlib import Path
from typing import Any

import yaml
from fastapi import APIRouter, HTTPException

router = APIRouter(prefix="/api/scenarios", tags=["scenarios"])

KNOWLEDGE_BASE_DIR = Path(__file__).resolve().parent.parent.parent / "knowledge_base" / "scenarios"


def _load_yaml(path: Path) -> dict[str, Any]:
    """Load a YAML file and return its contents."""
    if not path.exists():
        return {}
    return yaml.safe_load(path.read_text(encoding="utf-8")) or {}


def _extract_patient_info(scenario_dir: Path) -> dict[str, Any]:
    """Extract patient information from scenario YAML files."""
    identity = _load_yaml(scenario_dir / "identity_job.yaml")
    opening = _load_yaml(scenario_dir / "opening.yaml")
    symptoms = _load_yaml(scenario_dir / "symptoms.yaml")
    reason = _load_yaml(scenario_dir / "reason_wedding.yaml")
    emotional = _load_yaml(scenario_dir / "emotional_profile.yaml")
    
    # Parse identity content
    identity_content = identity.get("content", "")
    name = "Patient"
    age = 0
    occupation = ""
    
    if "Chloe Harrington" in identity_content:
        name = "Chloe Harrington"
    if "28" in identity_content:
        age = 28
    if "triathlete" in identity_content.lower():
        occupation = "Semi-professional Triathlete"
    
    # Get opening statement
    opening_content = opening.get("content", "").strip().strip('"')
    
    # Get case_id for scenario identification
    case_id = identity.get("case_id", scenario_dir.name)
    
    return {
        "id": scenario_dir.name,
        "case_id": case_id,
        "name": name,
        "age": age,
        "occupation": occupation,
        "gender": "Female" if "chloe" in name.lower() else "Unknown",
        "opening_statement": opening_content,
        "topic": identity.get("topic", ""),
    }


def _build_scenario(scenario_id: str, patient_info: dict) -> dict[str, Any]:
    """Build a complete scenario object for the UI."""
    return {
        "id": scenario_id,
        "groupName": "Amalgam Replacement",
        "title": "Pre-Wedding Aesthetic Consultation",
        "category": "Restorative Dentistry",
        "description": f"Patient {patient_info['name']} wants to replace silver fillings with white ones before their wedding.",
        "patientProfile": {
            "name": patient_info["name"],
            "age": patient_info["age"],
            "gender": patient_info["gender"],
            "occupation": patient_info["occupation"],
            "medicalHistory": "Healthy. No significant medical history.",
        },
        "brief": {
            "scenario": f"{patient_info['name']} is a {patient_info['age']}-year-old {patient_info['occupation'].lower()} who wants all silver (amalgam) fillings replaced with tooth-coloured (composite) fillings before their wedding next month.",
            "task": "Take a history, discuss treatment options, and address the patient's concerns about timing and aesthetics.",
            "instructions": [
                "You cannot examine the patient.",
                "Discuss the pros and cons of amalgam vs composite.",
                "Address any concerns about mercury or longevity.",
                "You have 8 minutes."
            ],
        },
        "goals": [
            "Establish rapport with the patient",
            "Take a focused dental history",
            "Explore the reason for the request",
            "Discuss treatment options clearly",
            "Address aesthetic and timing concerns"
        ],
        "clinicalNotes": "Patient is motivated by upcoming wedding. No symptoms reported."
    }


@router.get("")
async def list_scenarios() -> list[dict[str, Any]]:
    """List all available scenarios."""
    scenarios = []
    
    if not KNOWLEDGE_BASE_DIR.exists():
        return scenarios
    
    for scenario_dir in sorted(KNOWLEDGE_BASE_DIR.iterdir()):
        if scenario_dir.is_dir() and not scenario_dir.name.startswith("."):
            try:
                patient_info = _extract_patient_info(scenario_dir)
                scenario = _build_scenario(scenario_dir.name, patient_info)
                scenarios.append(scenario)
            except Exception as e:
                # Skip malformed scenarios
                continue
    
    return scenarios


@router.get("/{scenario_id}")
async def get_scenario(scenario_id: str) -> dict[str, Any]:
    """Get a specific scenario by ID."""
    scenario_dir = KNOWLEDGE_BASE_DIR / scenario_id
    
    if not scenario_dir.exists() or not scenario_dir.is_dir():
        raise HTTPException(status_code=404, detail=f"Scenario '{scenario_id}' not found")
    
    patient_info = _extract_patient_info(scenario_dir)
    return _build_scenario(scenario_id, patient_info)
